### PR TITLE
Discourage extending of functions in other packages for private use

### PR DESCRIPTION
**Pull request by rikhuijzer**

### CODE REVIEW

**Comment by KristofferC** [comment]
I don't agree with this. Extending methods is very "Julian" and should not be advised against.

**Comment by jpsamaroo** [comment]
Instead of discouraging this behavior (which is core to the language's extensibility), it would be more helpful to point out that it can cause invalidations which incur inference and compilation costs, but that those can be minimized with careful design (using SnoopCompile to reduce invalidations, especially).

**Comment by rikhuijzer** [comment]
I think that there has been a bit of a miscommunication which thankfully was cleared up by someone in a private chat.

In the first version of this PR, I forgot about the fact that extending methods can, of course, be very convenient and useful for downstream users. What I was aiming at was **private methods** on private types. So if a new type is created **inside** a module and is only meant to be used inside that module, then, I'd say that extending the **global** method table is not ideal. In many cases, it would then be more neat to define a local function which may avoid unnecessary invalidation, avoids modifying the global method table and even makes the code more readable since it's easier to match method calls with definitions

**Comment by Tokazama** [comment]
This is still very specific advice that is only true in specific situations. What if you have an internal array type? Should you not use the array interface? There are other solutions to invalidations too.

Perhaps this is good advice with a lot more context?

**Comment by KristofferC** [comment]
> I'm mostly trying to convey that extending methods from other packages is a global change which goes against modularity.

That's what the type piracy discussion is about.

> So if a new type is created inside a module and is only meant to be used inside that module, then, I'd say that extending the global method table is not ideal.

Why not? What if you want to use some interface?

I still think this is weird and should not be in the style guide. Where does this idea even come from? There are cases where extending methods can lead to bad effects (like https://github.com/invenia/Intervals.jl/issues/144) but I think that should be a very specific discussion and not general like this.

**Comment by rikhuijzer** [comment]
>> So if a new type is created inside a module and is only meant to be used inside that module, then, I'd say that extending the global method table is not ideal.
> Why not? 

Ah yes. I get it now. I cannot come up with any case where globally modifying the method table for a type under the control of one package can alter the program behavior of another package as long as the other package isn't using the type. So, it is global operation but as long as the type isn't used anywhere else it has no global effects.

> What if you want to use some interface?

Yeah sure. I was never arguing against those. 

> Where does this idea even come from? 

Things such as https://github.com/JuliaWeb/HTTP.jl/pull/845.

**Comment by Tokazama** [comment]
Just off the top of my head there are two situations where it causes problems to define a new method for a package's type in the way you are describing.

1. It isn't clear that the new method supports the purpose of the original function. Your example from HTTP.jl is one where it's not clear that there is a single `Int16` value that `FrameFlags` should convert into. A more obvious case would be using `+` to print to the REPL. 
2. It causes invalidations, which is what I think you're getting at here. The big problems with invalidation usually come when we run code that has poorly inferred components and then a new definition is added later that could change the outcome of one those poorly inferred methods in the method table. If there are invalidations because of a clearly typed conversion like `Int16(::MyType)`, then it's likely we need to address something upstream to fix this.

**Comment by mikmoore** [comment]
However, the "recreational" overloading of methods from other packages should be avoided. If your method is semantically the "same", such that code that uses the original function might end up being called on your new type, then dispatch away. But you should not overload a function for the name alone. For example,
```julia
struct Rectangle
    len::Float64
    wid::Float64
end
Base.length(x::Rectangle) = x.len # not at all what `Base.length` is normally used for
```
is probably bad practice and a new (not `Base.`) function should be defined instead. However, this isn't piracy and this method is virtually harmless as far as I know -- it's just bad practice and semantic confusion.

But i don't think this pattern is common enough for us to address. To discuss it would probably cause as much harm as good.

**Comment by Tokazama** [comment]
What you're describing was the first situation I mentioned. It's fine to clarify this further but any discussion about the method table isn't related to that point.

**Comment by rikhuijzer** [comment]
FWIW, it is possible to affect LLVM code outside the current package (but without affecting program behavior) since the compiler will in some use the method table for type inference.

```julia
module A

foo(x::Int) = x::Int
f(x) = foo(Base.inferencebarrier(x))

end
```
```julia
julia> @code_warntype A.f(1)
MethodInstance for Main.A.f(::Int64)
  from f(x) in Main.A at /home/rik/git/tmp.jl:5
Arguments
  #self#::Core.Const(Main.A.f)
  x::Int64
Body::Int64
1 ─ %1 = Base.getproperty(Main.A.Base, :inferencebarrier)::Core.Const(Base.inferencebarrier)
│   %2 = (%1)(x)::Any
│   %3 = Main.A.foo(%2)::Int64
└──      return %3
```
versus
```julia
module A

foo(x::Int) = x::Int
f(x) = foo(Base.inferencebarrier(x))

end

module B
    import ..A

    struct T end
    A.foo(x::T) = string(x)::String
end
``` 
```julia
julia> @code_warntype A.f(1)
MethodInstance for Main.A.f(::Int64)
  from f(x) in Main.A at /home/rik/git/tmp.jl:5
Arguments
  #self#::Core.Const(Main.A.f)
  x::Int64
Body::Union{Int64, String}
1 ─ %1 = Base.getproperty(Main.A.Base, :inferencebarrier)::Core.Const(Base.inferencebarrier)
│   %2 = (%1)(x)::Any
│   %3 = Main.A.foo(%2)::Union{Int64, String}
└──      return %3
```
This shows that "harmless" non-piracy can still affect the code outside of the current module. 

Also, I think that extending methods from other packages can sometimes be quite hard to follow for developers. If it is reasonable to give the method an clear and unique name, then that can be much easier to follow.

So, this is not just about method invalidations as people keep bringing up; it is about affecting global state and readability as well.

Suggestions to improve the wording of this PR are welcome

**Comment by DilumAluthge** [comment]
I'm also opposed to this pull request - my reasoning is pretty much the same as Kristoffer's.

**Comment by Tokazama** [comment]
@ikhuijzer, using your example, if `foo` is meant to be an extensible part of a public API then it is poorly designed. That's an issue with module `A`, not `B`. It's also fine to have a single invalidation here and there. The issue is when you have a widely used method like `length` and it can result in hundreds of invalidations. We don't often have that issue because it's usually easy to infer the result of `length`.

**Comment by rikhuijzer** [comment]
> What version are you on?

Julia 1.8-rc1.

**Comment by JeffBezanson** [comment]
I think I see what you are getting at; the statement "it is better to create a local function than to add a method **if it is easy to do so**" is basically true. It is somewhat like marking things `static` or `private`. The "easy" part is doing a lot of work: if that transformation is easy, you are not using OO or generic programming in that part of your code.

Sometimes there is a heightened sense of "danger" from adding a method in julia as opposed to class-based languages, due to some awareness of how the system works (i.e. mutating a public method table). But really the situation is not so different from e.g. python, where any time you write a method called `x` you are potentially extending every `y.x()` call site in the system.
